python/llm/example/GPU/Pipeline-Parallel-FastAPI/pipeline_models.py

```diff
@@ -0,0 +1,304 @@
+
+        for idx in range(self.num_layers):
+            decoder_layer = self.layers[self.layer_start + idx]
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+
+            past_key_value = past_key_values[idx] if past_key_values is not None else None
+
+            layer_outputs = decoder_layer(
+                hidden_states,
+                attention_mask=attention_mask,
+                position_ids=position_ids,
+                past_key_value=past_key_value,
+                output_attentions=output_attentions,
+                use_cache=use_cache,
+            )
+
+            hidden_states = layer_outputs[0]
+
+            if use_cache:
+                next_decoder_cache += (layer_outputs[2 if output_attentions else 1],)
+
+            if output_attentions:
+                all_self_attns += (layer_outputs[1],)
+        
+        if self.pp_config.is_tail:
+            hidden_states = self.norm(hidden_states)
+
+        # add hidden states from the last decoder layer
+        if output_hidden_states:
+            all_hidden_states += (hidden_states,)
+
+        next_cache = next_decoder_cache if use_cache else None
+        if not return_dict:
+            return tuple(v for v in [hidden_states, next_cache, all_hidden_states, all_self_attns] if v is not None)
+        return BaseModelOutputWithPast(
+            last_hidden_state=hidden_states,
+            past_key_values=next_cache,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attns,
+        )
+
+
+def load_model(checkpoint):
+    from llama_models import LlamaForCausalLM
+    if 'llama' in checkpoint.lower():
+        model = LlamaForCausalLM.from_pretrained(checkpoint, low_cpu_mem_usage=True, torch_dtype=torch.float16)
+    return model
+
+class ModelRunner:
+    
+    def __init__(self, checkpoint, rank, world_size, low_bit):
+        import intel_extension_for_pytorch as ipex
+        import sys
+        self.pp_config = PPConfig(rank, world_size)
+        
+        start = time.perf_counter()
+        model = load_model(checkpoint)
+        end = time.perf_counter()
+        logger.info(f"Time to load weights: {end - start:.2f}s")
+        from ipex_llm import optimize_model
+
+        model = optimize_model(model, low_bit=low_bit)
+        
+        model = model.to(torch.float32).to(f'xpu:{rank}')
+        self.model = model
+        self.rank = rank
+        self.world_size = world_size
+        self.pre_rank = (self.rank - 1) % self.world_size
+        self.next_rank = (self.rank + 1) % self.world_size
+        self.hidden_size = self.model.config.hidden_size
+    
+                
+    def generate(self, input_ids=None, max_tokens=5, attention_mask=None):
+        times = []
+        with torch.no_grad():
+            _input_ids = None
+            _past_key_values = None
+            bs = input_ids.shape[0]
+            output_ids = input_ids.clone()
+            for i in range(max_tokens):
```

Comment:
I don't think this is truly pipeline parallel. 

Let's take the first half of the model for example assuming PP=2, the model takes `input_ids` and run forward on line 268, and it then blocking waits on line 282 for the next token while the 2nd part of the model is running to generate the next token. 

When `generate` finished generating all the tokens, it starts to work on the next request, so the GPU is idle half of the time.